src/stream/src/executor/backfill.rs

```diff
@@ -0,0 +1,380 @@
+                                for chunk in upstream_chunk_buffer.drain(..) {
+                                    yield Message::Chunk(Self::mapping_chunk(
+                                        chunk,
+                                        &upstream_indices,
+                                    ));
+                                }
+
+                                // Finish backfill.
+                                break 'backfill_loop;
+                            }
+                            Some(chunk) => {
+                                // Raise the current position.
+                                // As snapshot read streams are ordered by pk, so we can
+                                // just use the last row to update `current_pos`.
+                                self.current_pos = Some(
+                                    chunk
+                                        .rows()
+                                        .last()
+                                        .unwrap()
+                                        .1
+                                        .row_by_indices(&table_pk_indices),
+                                );
+
+                                yield Message::Chunk(Self::mapping_chunk(
+                                    chunk,
+                                    &self.upstream_indices,
+                                ));
+                            }
+                        }
+                    }
+                }
+            }
+        }
+
+        let mut finish_on_barrier = |msg: &Message| {
+            if let Some(barrier) = msg.as_barrier() {
+                self.progress.finish(barrier.epoch.curr);
+            }
+        };
+
+        tracing::debug!(
+            actor = self.actor_id,
+            "Backfill has already finished and forward messages directly to the downstream"
+        );
+
+        // Backfill has already finished.
+        // Forward messages directly to the downstream.
+        let upstream = upstream
+            .map(move |result| result.map(|msg| Self::mapping_message(msg, &upstream_indices)));
+        #[for_await]
+        for msg in upstream {
+            let msg: Message = msg?;
+            finish_on_barrier(&msg);
+            yield msg;
+        }
+    }
+
+    #[expect(clippy::needless_lifetimes, reason = "code generated by try_stream")]
+    #[try_stream(ok = Option<StreamChunk>, error = StreamExecutorError)]
+    async fn snapshot_read(table: &StorageTable<S>, epoch: u64, current_pos: Option<Row>) {
+        // `current_pos` is None means it needs to scan from the beginning, so we use Unbounded to
+        // scan. Otherwise, use Excluded.
+        let range_bounds = if let Some(current_pos) = current_pos {
+            (Bound::Excluded(current_pos), Bound::Unbounded)
+        } else {
+            (Bound::Unbounded, Bound::Unbounded)
+        };
+        let iter = table
+            .batch_iter_with_pk_bounds(
```

Comment:
This will wait for an uncommitted epoch under the current architecture. It seems that after a checkpoint, we buffer all chunks from the upstream and also make no progress on the snapshot side in a few following epochs.

I'm considering whether this will lead to performance issues. What's worse, if `checkpoint_frequency = 1` and with the concurrent checkpoint on, it's possible that a new checkpoint always comes with the previous one uncommitted, then it seems we will be in a live lock. 🤔

Reply:
The `BackfillExecutor` sends the checkpoint barrier to the downstream, before read it (or wait epoch), so I think it wouldn't be in a live lock.

Reply:
Of course, it will increase the barrier latency while waiting committed data. We are planning to use read uncommitted data instead of reading committed data in the future (after the PR of local state storage has been merged).

Reply:
But we can't make any progress on the snapshot side in this case. 😨

Reply:
Actually, we can support read uncommitted directly by changing `HummockReadEpoch::Committed` to `HummockReadEpoch::NoWait`, since we already schedule the `ChainNode` with the upstream mv in the same worker. In this way we can process every barrier in a same way.  I am just wondering whether to support uncommitted read directly in this PR.